docs/architecture/architecture.md

```diff
@@ -0,0 +1,671 @@
+  instances and the `cc-shim` ones.
+- Route the I/O streams and signals between the various `cc-shim` instances and
+  the `cc-agent`.
+
+`cc-proxy` API is available through a single socket for all `cc-shim` and `cc-runtime`
+instances to connect to. `cc-proxy` can be configured to use a UNIX or a TCP
+socket, and by default will handle connection over a UNIX socket.
+
+The protocol on the `cc-proxy` socket supports the following commands:
+- `RegisterVM`: Used by `cc-runtime` to let `cc-proxy` know about a newly created
+  virtual machine. When registering a new VM, `cc-runtime` will ask `cc-proxy`
+  for a token and pass that token to the `cc-shim` instance it will eventually
+  create for handling the container process inside that VM.
+- `UnregisterVM`: Does the opposite of what `RegisterVM` does, indicating to
+  the proxy it should release resources created by `RegisterVM`.
+- `AttachVM`: It can be used to associate clients to an already known VM.
+  Optionally `AttachVM` senders can ask `cc-proxy` for a token. In the OCI `exec`
+  command case, this token will be used by the `cc-shim` instance that monitors
+  the executed process inside the container.
+- `Hyper` : This payload will forward an hyperstart command to the `cc-agent`.
+- `ConnectShim`: Used by `cc-shim` instances to connect to `cc-proxy` and let
+  it know which container process it wants to monitor. `cc-shim` sends the token
+  they've been given by `cc-runtime` as part of the `ConnectShim` payload. With
+  that information, `cc-proxy` builds its I/O and signal multiplexing routes
+  between the VM process containers and the host `cc-shim` instances.
+- `DisconnectShim`: Used by `cc-shim` instances to disconnect themselves from
+  `cc-proxy` and let it know that they stop monitoring their container process.
+
+For more details about `cc-proxy`'s protocol, theory of operations or debugging
+tips, please read
+[`cc-proxy` README](https://github.com/clearcontainers/proxy) or the
+[proxy api `godoc`](https://godoc.org/github.com/clearcontainers/proxy/api).
+
+## Shim
+
+A container process reaper, such as Docker's `containerd-shim` or crio's `conmon`,
+is designed around the assumption that it can monitor and reap the actual container
+process. As the container process reaper runs on the host, it cannot directly
+monitor a process running within a virtual machine. At most it can see the QEMU
+process, but that is not enough. With Clear Containers, `cc-shim` acts as the
+container process that the container process reaper can monitor. Therefore
+`cc-shim` needs to handle all container I/O streams (`stdout`, `stdin` and `stderr`)
+and forward all signals the container process reaper decides to send to the container
+process.
+
+`cc-shim` has an implicit knowledge about which VM agent will handle those streams
+and signals and thus acts as an encapsulation layer between the container process
+reaper and the `cc-agent`. `cc-shim`:
```

Comment:
Maybe put `cc-shim`: on a new paragraph, and expand a little to something like 'cc-shim performs the following steps:' or similar.